lib/sotest.nix

```diff
@@ -3,30 +3,94 @@
 with pkgs.lib;
 
 rec {
-  projectConfigJSON = {
-      boot_prerequisites ? [],
-      extra_dependencies ? [],
-      boot_panic_patterns ? [],
-      boot_items ? []
+  # Create an attribute set of a sotest test run with the format specified in
+  # https://docs.sotest.io/project_config
+  # In order to upload it to a sotest instance, it should be converted to
+  # JSON or YAML, see `asJSONFile` and `asYAMLFile`.
+  projectConfigJSON =
+    { boot_prerequisites ? []
+    , extra_dependencies ? []
+    , boot_panic_patterns ? []
+    , boot_items ? []
     }: builtins.toJSON {
       inherit
         boot_items
         boot_panic_patterns
         boot_prerequisites
-        extra_dependencies;
+        extra_dependencies
+        ;
     };
+
+  # Create a pretty-printed JSON file from the input
   asJSONFile = jsonContent: pkgs.runCommandNoCC "content.json"
     { nativeBuildInputs = [ pkgs.jq ]; }
     "jq '.' ${pkgs.writeText "content.json" jsonContent} > $out";
+
+  # Create a pretty-printed YAML file from the input
   asYAMLFile = jsonContent: pkgs.runCommandNoCC "content.yaml"
     { nativeBuildInputs = [ pkgs.yq ]; }
     "yq -y '.' ${pkgs.writeText "content.yaml" jsonContent} > $out";
-  linuxBootItem = {
-      boot_item_timeout ? 300,
-      name,
-      kernel,
-      kernelParams ? [ "console=@{linux_terminal}" ],
-      initrd
+
+  # Calculate list of direct dependencies of a derivation
+  flatReferences = drv: pkgs.stdenv.mkDerivation {
+    name = "flat-closure-info";
+    __structuredAttrs = true;
+    exportReferencesGraph.closure = [ drv ];
+    preferLocalBuild = true;
+    PATH = "${pkgs.coreutils}/bin:${pkgs.jq}/bin";
+    builder = builtins.toFile "builder" ''
+      . .attrs.sh
+
+      out=''${outputs[out]}
+
+      jq -r '.exportReferencesGraph.closure[0] as $x | .closure[] | select(.path == $x) | .references[]' \
+        < .attrs.json \
+        > $out
+    '';
+  };
```

Comment:
Why can't this just be a call to `pkgs.runCommandNoCC`, as it is done, e.g. in zipBundleFromTestrunClosure?

Reply:
The `exportReferencesGraph.closure = [ drv ];` does some hardly documented magic to the derivation, so the default builder script (which is in use when doing `runCommandNoCC`) doesn't find `/setup` any longer... so the other ways are also messier.

Reply:
Can we then at least document why we're doing it like this (e.g. in a comment) while we still know the reason?